File: Go/dataStructures/linkedQueue.go
package datastructures

type node struct {
	Data any
	Next *node
}

type Queue struct {
	head   *node
	tail   *node
	length int
}

func (ll *Queue) enqueue(n any) {
	var newNode node
	newNode.Data = n

	if ll.tail != nil {
		ll.tail.Next = &newNode
	}
	ll.tail = &newNode
	if ll.head == nil {
		ll.head = &newNode
	}
	ll.length++
}

func (ll *Queue) dequeue() any {
	if ll.isEmpty() {
		return -1
	}
	data := ll.head.Next
	ll.head = ll.head.Next
	if ll.head == nil {
		ll.tail = nil
	}
	ll.length--
	return data
}

// isEmpty it will check our list is empty or not
func (ll *Queue) isEmpty() bool {
	return ll.length == 0
}

// len is return the length of queue
func (ll *Queue) len() int {
	return ll.length
}

// frontQueue it will return the front data
func (ll *Queue) frontQueue() any {
	return ll.head.Data
}

// backQueue it will return the back data
func (ll *Queue) backQueue() any {
	return ll.tail.Data
}

File: TypeScript/data Structures/LinkedQueue.ts

import { assertEquals } from "https://deno.land/std@0.192.0/testing/asserts.ts";

type NodeQueue<T> = {
  value: T,
  next?: NodeQueue<T>
}


export class LinkedQueue<T>{
  public size: number;
  public head?: NodeQueue<T>;
  private tail?: NodeQueue<T>;

  constructor() {
    this.head = this.tail = undefined;
    this.size = 0;
  }
  enqueue(item: T): void {
    const node = { value: item } as NodeQueue<T>;
    this.size++

    if (!this.tail) {
      this.tail = this.head = node;
      return
    }
    this.tail.next = node
    this.tail = node
  }
  dequeue(): T | undefined {
    if (!this.head) throw new Error("Queue UnderFlow")
    this.size--
    const head = this.head;
    this.head = this.head.next
    return head.value
  }


  /**
  * Returns the item at the front of the queue.
  *
  * @returns The item at the front of the queue or null if the queue is empty.
  */
  peek(): T | undefined | null {

    if (this.isEmpty()) {
      return null;
    }
    return this.head?.value;
  }

  /**
   * Checks if the queue is empty.
   *
   * @returns {boolean} Whether the queue is empty or not.
   */
  isEmpty(): boolean {
    return this.size === 0
  }

  /**
   * Returns the number of items in the queue.
   *
   * @returns {number} The number of items in the queue.
   */
  length(): number {
    return this.size;
  }
}


Deno.test(() => {
  const queue = new LinkedQueue()
  queue.enqueue("hola")
  assertEquals(queue.length() === 1)

})
